docs/site/Req-res-cycle.md

```diff
@@ -0,0 +1,216 @@
+
+[Services](https://loopback.io/doc/en/lb4/Services.html) are LoopBack's helper
+units with support for dependency injection. They come in three varieties -
+proxy, class, and provider.
+
+A proxy service acts as a proxy to an external service. A class service creates
+an instance of a helper class. A provider service resolves a value.
+
+All of them have access to the app, request, and response objects via dependency
+injections, and can influence the handling of the request-response cycle.
+
+Services can be injected in the controller constructor so controller methods may
+use them.
+
+##### Repositories
+
+[Repositories](https://loopback.io/doc/en/lb4/Repositories.html) are the links
+between controllers and the data. They use an underlying datasource and
+a connector to interact with the data.
+
+Controller methods can call corresponding methods in the respository to read
+from or write to the database.
+
+Repositories can be injected in the controller constructor so controller methods
+may use them.
+
+##### Interceptors
+
+[Interceptors](https://loopback.io/doc/en/lb4/Interceptors.html) intercept
+invocation of method on classes decorate with the `@intercept()` decorator.
+
+All controllers, services, and repositories; and their methods can be intercepted.
+This makes interceptors a powerful participant in the request-response cycle,
+since they can modify the request and response objects and call their methods.
+
+##### Walk through of a request
+
+Now that we know all the components that may be involve in a request to a
+controller endpoint, let's walk through a request to an endpoint - `/ping`
+endpoint of a scaffolded LoopBack app.
+
+The `PingController` controller has a method named `ping()`, decorated with a
+`get()` decorator. This setup creates the `/ping` endpoint on the app.
+
+The app is configured to use the `MySequence` sequence.
+
+Now this is a step-wise sequence of what happens when a request is made to
+`http://localhost:4000/ping`.
+
+1. The request is intercepted by `MySequence`.
+2. The route is then identified by `FindRoute`. There is a handler for this request.
+3. `ParseParams` then tries to parses any parameters might have been submitted.
+There is none.
+4. `InvokeMethod` then invokes the handler with empty parameters. It returns
+whatever the handler returns, in this case, it is JSON object with `greeting`,
+`date`, `url`, and `headers` properties.
+5. `Send` then sends this JSON object back to the client.
+
+#### Request to a non-controller endpoint
+
+Non-controller endpoints can be created by static files, redirection rules,
+custom Express routes, and non-controller endpoints added by components.
+
+Requests to non-controller endpoints are also intercepted by the app's sequence,
+but the control is handed over to the underlying Express middleware from the
+`InvokeMethod` phase, and never make it to `Send` unless the middleware calls
+`next()`. Unhandled error from Express middleware still reach `Reject`.
```

Comment:
I like how we show the differences between these two endpoints. But I am a bit confused about `services` and the `interceptor` part. From my understanding, the structure we show above is:

```
             / ==> controller endpoint -> Services, Interceptors -> repositories
Sequence ---
             \ ==> non-controller endpoint
```

But IIUC I think the interceptors can also be use with non-controller endpoints. Maybe we make it more clear here.

Reply:
Can you point out where this structure is contradicted?

```
             / ==> controller endpoint -> Services, Interceptors -> repositories
Sequence ---
             \ ==> non-controller endpoint
```

Reply:
It looks like the interceptors cannot be used in non-controller endpoints. Is it true?

Reply:
Interceptors work only on Controllers and Repositories, so non-controller endpoint handlers can't be intercepted.

Reply:
![image](https://user-images.githubusercontent.com/50331796/77331406-0e3fb700-6cf7-11ea-92f6-42a7a58c4609.png)
I think I misunderstood the hander in the diagram as the non-controller endpoints.

Reply:
Works on Services too.

Reply:
IIRC, global interceptors also intercept non-controller routes.